src/library-authoring/components/LibraryComponents.tsx

```diff
@@ -0,0 +1,141 @@
+import React, { useEffect, useMemo } from 'react';
+
+import { CardGrid } from '@openedx/paragon';
+import { NoComponents, NoSearchResults } from '../EmptyStates';
+import { useLibraryBlockTypes, useLibraryComponentCount, useLibraryComponents } from '../data/apiHook';
+import { ComponentCard, ComponentCardLoading } from './ComponentCard';
+
+type LibraryComponentsProps = {
+  libraryId: string,
+  filter: {
+    searchKeywords: string,
+  },
+  variant: string,
+};
+
+/**
+ * Library Components to show components grid
+ *
+ * Use style to:
+ *   - 'full': Show all components with Infinite scroll pagination.
+ *   - 'preview': Show first 4 components without pagination.
+ */
+const LibraryComponents = ({
+  libraryId,
+  filter: { searchKeywords },
+  variant,
+}: LibraryComponentsProps) => {
+  const { componentCount } = useLibraryComponentCount(libraryId, searchKeywords);
+  const {
+    hits,
+    isFetching,
+    isFetchingNextPage,
+    hasNextPage,
+    fetchNextPage,
+  } = useLibraryComponents(libraryId, searchKeywords);
+
+  const { componentList, tagCounts } = useMemo(() => {
+    const result = variant === 'preview' ? hits.slice(0, 4) : hits;
+    const tagsCountsResult = {};
+    result.forEach((component) => {
+      if (!component.tags) {
+        tagsCountsResult[component.id] = 0;
+      } else {
+        tagsCountsResult[component.id] = (component.tags.level0?.length || 0)
+              + (component.tags.level1?.length || 0)
+              + (component.tags.level2?.length || 0)
+              + (component.tags.level3?.length || 0);
+      }
+    });
+    return {
+      componentList: result,
+      tagCounts: tagsCountsResult,
+    };
+  }, [hits]);
+
+  // TODO add this to LibraryContext
+  const { data: blockTypesData } = useLibraryBlockTypes(libraryId);
+  const blockTypes = useMemo(() => {
+    const result = {};
+    if (blockTypesData) {
+      blockTypesData.forEach(blockType => {
+        result[blockType.blockType] = blockType;
+      });
+    }
+    return result;
+  }, [blockTypesData]);
+
+  const { showLoading, showContent } = useMemo(() => {
+    let resultShowLoading = false;
+    let resultShowContent = false;
+
+    if (isFetching && !isFetchingNextPage) {
+      // First load; show loading but not content.
+      resultShowLoading = true;
+      resultShowContent = false;
+    } else if (isFetchingNextPage) {
+      // Load next page; show content and loading.
+      resultShowLoading = true;
+      resultShowContent = true;
+    } else if (!isFetching && !isFetchingNextPage) {
+      // State without loads; show content.
+      resultShowLoading = false;
+      resultShowContent = true;
+    }
+    return {
+      showLoading: resultShowLoading,
+      showContent: resultShowContent,
+    };
+  }, [isFetching, isFetchingNextPage]);
```

Comment:
You don't need this logic here. React Query already provides you everything you need. In fact, putting this in `useMemo` is causing the results to flicker every time the user types a letter in the search result. I'll show you a separate suggestion below of how to get the exact same result without any new variables.

https://github.com/user-attachments/assets/b239dce8-cd17-4fbe-918f-1905d68cf7b5


```suggestion
```

Reply:
Loading deleted: https://github.com/openedx/frontend-app-course-authoring/pull/1109/commits/4fe14c03ac60dd0d3c588e71ee8e784ebc8e15f5